src/Adapter/Order/OrderAmountUpdater.php

```diff
@@ -225,66 +224,70 @@ private function updateOrderCartRules(
-        $carrierId = $order->id_carrier;
-        $totalMethod = ($firstInvoice === null || $firstInvoice->id == $invoice->id) ? Cart::BOTH : Cart::BOTH_WITHOUT_SHIPPING;
-        $invoice->total_paid_tax_excl = Tools::ps_round(
-            (float) $cart->getOrderTotal(false, $totalMethod, $invoiceProducts, $carrierId),
-            $computingPrecision
-        );
-        $invoice->total_paid_tax_incl = Tools::ps_round(
-            (float) $cart->getOrderTotal(true, $totalMethod, $invoiceProducts, $carrierId),
-            $computingPrecision
-        );
+        foreach ($invoiceCollection as $invoice) {
+            // If all the invoice's products have been removed the offset won't exist
+            $currentInvoiceProducts = isset($invoiceProducts[$invoice->id]) ? $invoiceProducts[$invoice->id] : [];
 
-        $invoice->total_products = (float) $cart->getOrderTotal(
-            false,
-            Cart::ONLY_PRODUCTS,
-            $invoiceProducts,
-            $carrierId
-        );
-        $invoice->total_products_wt = (float) $cart->getOrderTotal(
-            true,
-            Cart::ONLY_PRODUCTS,
-            $invoiceProducts,
-            $carrierId
-        );
+            // Shipping are computed on first invoice only
+            $carrierId = $order->id_carrier;
+            $totalMethod = ($firstInvoice === null || $firstInvoice->id == $invoice->id) ? Cart::BOTH : Cart::BOTH_WITHOUT_SHIPPING;
+            $invoice->total_paid_tax_excl = Tools::ps_round(
+                (float) $cart->getOrderTotal(false, $totalMethod, $currentInvoiceProducts, $carrierId),
+                $computingPrecision
+            );
+            $invoice->total_paid_tax_incl = Tools::ps_round(
+                (float) $cart->getOrderTotal(true, $totalMethod, $currentInvoiceProducts, $carrierId),
+                $computingPrecision
+            );
 
-        $invoice->total_discount_tax_excl = $invoice->total_discount_tax_incl = 0;
-        foreach ($order->getCartRules() as $orderCartRuleData) {
-            $orderCartRule = new OrderCartRule($orderCartRuleData['id_order_cart_rule']);
-            if ($orderCartRule->id_order_invoice == 0 || $orderCartRule->id_order_invoice == $invoice->id) {
-                $invoice->total_discount_tax_incl += $orderCartRule->value;
-                $invoice->total_discount_tax_excl += $orderCartRule->value_tax_excl;
+            $invoice->total_products = (float) $cart->getOrderTotal(
+                false,
+                Cart::ONLY_PRODUCTS,
+                $currentInvoiceProducts,
+                $carrierId
+            );
+            $invoice->total_products_wt = (float) $cart->getOrderTotal(
+                true,
+                Cart::ONLY_PRODUCTS,
+                $currentInvoiceProducts,
+                $carrierId
+            );
+
+            $invoice->total_discount_tax_excl = $invoice->total_discount_tax_incl = 0;
+            foreach ($order->getCartRules() as $orderCartRuleData) {
```

Comment:
Why not using `$cart->getOrderTotal(true, Cart::ONLY_DISCOUNTS, $currentInvoiceProducts, $carrierId)` here?

Reply:
This code existed prior to this PR (but I'm not sure to understand what you want to replace with this call to $cart->getOrderTotal ?)

Reply:
It looks like you're manually iterating through all cart rules to get the total discount for the invoice, I was just suggesting using `getOrderTotal()` with `Cart::ONLY_DISCOUNTS` to simplify but maybe I'm missing something?

Reply:
No I think you're right I didn't think about it but it's worth trying it! It tests are still green it's a go! 👍

Reply:
damn my message was erased, yes I just understood you were talking about the code below your comment.
Hm, maybe @jolelievre knows if there's a limitation keeping us from using `getOrderTotal()` with `Cart::ONLY_DISCOUNTS` ?

(it only iterates through cart rules applied to this order hopefully :D )

Reply:
Thinking about it, I'm not sure getOrderTotal() can calculate discounts only for a specific invoice inside an order, this is probably why @jolelievre did it this way, can you confirm ?

Reply:
damn my github is so buggy, messages disappear, and I see new messages much later even after refresh 😓